### PR TITLE
Properly resolve cupertino theme data defaults for MaterialbasedCupertinoThemeData

### LINKED ISSUE #51899: CupertinoBar color does not update with brightness, for dark mode.
## Steps to Reproduce

1. Run `flutter create bug`.
2. Update the main.dart as follows: 

```
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        cupertinoOverrideTheme:
            const CupertinoThemeData(brightness: Brightness.dark),
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text(
                'You have pushed the button this many times:',
              ),
              Text(
                '$_counter',
                style: Theme.of(context).textTheme.display1,
              ),
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: _incrementCounter,
          tooltip: 'Increment',
          child: Icon(Icons.add),
        ),
        bottomNavigationBar: CupertinoTabBar(
          items: [
            BottomNavigationBarItem(
              icon: Icon(Icons.ac_unit),
              title: Text('AC'),
            ),
            BottomNavigationBarItem(
              icon: Icon(Icons.access_alarm),
              title: Text('Alarm'),
            ),
          ],
        )
        );
  }
}
```
**Expected results:** 

 If brightness is Light, the status bar icons should be black or viceversa. If brightness is Black, app bar should be black.
```
cupertinoOverrideTheme:
            const CupertinoThemeData(brightness: Brightness.dark), 
```
**Actual results:** 
Only the CupertinoTabBar is updating if overriding, if not overriding but dark theme is on it works too but no the app bar.

image

<details>
  <summary>Logs</summary>

Flutter Doctor
```
[✓] Flutter (Channel stable, v1.12.13+hotfix.8, on Mac OS X 10.15.3 19D76, locale en-CO)
    • Flutter version 1.12.13+hotfix.8 at /Users/nicolas.villamizar/flutter
    • Framework revision 0b8abb4724 (3 weeks ago), 2020-02-11 11:44:36 -0800
    • Engine revision e1e6ced81d
    • Dart version 2.7.0

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
    • Android SDK at /Users/nicolas.villamizar/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    • CocoaPods version 1.8.4

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 43.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.42.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.8.1

[✓] Connected device (1 available)
    • iPhone 11 • A27B9117-A941-4F9F-8177-60DEA19279F7 • ios • com.apple.CoreSimulator.SimRuntime.iOS-13-3 (simulator)

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

Captures the provided `BuildContext` so that the colors can be resolved lazily at a later point in time. Currently `CupertinoThemeData` eagerly resolves all its attributes in `CupertinoTheme.of`, causing the calling widget (i.e., `BuildContext`) to declare dependencies on more `InheritedWidget`s than necessary. 
It could be problematic if a `_CupertinoDynamicThemeData` is passed to a descendant widget or render objects. 

## Related Issues

Fixes https://github.com/flutter/flutter/issues/51899

## Tests

- Updated "Default theme has defaults" to cover `barBackgroundColor`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter/lib/src/material/theme_data.dart [hunk 1]
```diff
@@ -1576,6 +1576,9 @@ class MaterialBasedCupertinoThemeData extends CupertinoThemeData {
   @override
   Color get scaffoldBackgroundColor => _cupertinoOverrideTheme.scaffoldBackgroundColor ?? _materialTheme.scaffoldBackgroundColor;
 
+  @override
+  Color get barBackgroundColor => _cupertinoOverrideTheme.barBackgroundColor ?? _materialTheme.bottomAppBarTheme.color ?? _materialTheme.bottomAppBarColor;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/theme_data.dart]: I'm not so sure `bottomAppBarColor` is the correct color that we want to derive `barBackgroundColor` from.

**Comment by xster** [comment]
I vaguely remember that this may have been intentional since there's no good fit for a muted color that's a good match from the material theme colors. 

Are there attributes in the theme data that's an "alternate" or "muted" bar color?

File: packages/flutter/lib/src/cupertino/theme.dart [hunk 2]
```diff
@@ -303,6 +305,80 @@ class CupertinoThemeData extends NoDefaultCupertinoThemeData with Diagnosticable
...
+  }) : _themeData = baseThemeData,
+       _context = context;
+
+  @override
+  Brightness? get brightness => _themeData.brightness;
+
+  @override
+  _CupertinoThemeDefaults get _defaults => _themeData._defaults;
+  final CupertinoThemeData _themeData;
+  final BuildContext _context;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/theme.dart:321]: I tried to put some asserts around the `_context` getter to make sure it's only accessed during persistent callback phase, but that broke some of our tests (those retrieve the theme data and put them in `expect`).

**Comment by goderbauer** [comment]
@LongCatIsLooong Do you still have plans for this very old PRs or should it be closed?

**Comment by LongCatIsLooong** [comment]
@xster @HansMuller does this look like a reasonable approach to you?

**Comment by HansMuller** [comment]
I have reviewed this carefully however we don't typically build classes like _CupertinoDynamicThemeData that hang onto BuildContext values.

**Pull request closed without merging**